### PR TITLE
C#: Reimplement flow-summary compilation

### PR DESCRIPTION
**Pull request by hvitved**
### Overview

When flow summaries (aka. `LibraryTypeDataFlow`) were first introduced, they were limited to simple summaries such as taint-propagation from `s` to `int.Parse(s)`. Such summaries were compiled down to simple local flow steps, e.g. a taint-step from `s` to `int.Parse(s)`.

When flow summaries were [later generalized to include content-sensitivity](https://github.com/github/codeql/pull/3270), flow summaries were compiled down to [state-machines](https://github.com/github/codeql/pull/3270/files?file-filters%5B%5D=.qll#diff-623b51985c523c1ec3d03ed19b4ef882R417-R422), which would break up summaries into atomic reads, stores, and local steps. However, the compilation would still happen _at each call site_, which meant that we would effectively inline the summarized methods.

This PR removes inlining by instead modelling summarized methods as normal methods. That is, we now synthesize data-flow nodes for parameters, internal nodes, and returns, and let the shared data-flow library take care of matching call sites against returns, just like it does for non-summarized methods. An additional benefit from this change is that flow through delegates passed to summarized methods, such as `Foo` in `ints.Select(Foo)`, now get a more accurate modelling; we synthesize a call to the delegate for parameter `p2` inside `Select` (`call_p2`), an argument node of that delegate call (`arg`), as well as an output node (`out_p2`):
```
Select(this IEnumerable<T> p1, Func<T, S> p2)
                           /
			  /
                      read([])
          	        /
		       /
	              arg --(argument 0)-- call_p2 --(output)-- out_p2
	                                                        /
			                     	               /
                      			                   store([])
		                     	                     /
                     			                    /
                     			                 return  
```
The vertical steps above are read/store steps, and the horizontal steps are input/output steps.

### Implementation

Since we are no longer inlining the compiled steps, we can no longer perform call-site based [access path adjustments](https://github.com/github/codeql/pull/3994), and that logic has therefore been removed.

The following new entities are introduced:
- `SummaryDelegateCall`: This models calls to delegates inside summarized methods (`call_p2` above).
- `SummaryParameterNode`: This models parameters of summarized methods (`p1` and `p2` above).
- `SummaryInternalNode`: This models the internal steps inside summarized methods, and replaces the old `LibraryCodeNode`.
- `SummaryReturnNode`: This models returns out of summarized methods (`return` above).
- `SummaryDelegateOutNode`: This models the output of calling delegates inside summarized methods (`out_p2` above).
- `SummaryDelegateArgumentNode`: This models arguments to delegate calls inside summarized methods (`arg` above).

The entities `ImplicitDelegateCall`, `ImplicitDelegateOutNode`, and `ImplicitDelegateArgumentNode` have been removed (they used to model calls to delegates inside summarized methods, but the modelled call happened at the call-sites).

### Charp-Differences

https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/513/

This rewrite incidentally uncovered a bug in the old implementation, when arguments are passed to a summarized `params` method. In order to better compare differences, the first commit addresses that bug.

On `mono`, some results are missing: It turns out that this is because of the [buggy `fieldFlowBranchLimit` implementation](https://github.com/github/codeql/pull/3071), where some flow is now filtered away, because summarized method calls are treated like normal method calls.

### CODE REVIEW

File: csharp/ql/src/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll [hunk 1]
```diff
@@ -1414,662 +1514,353 @@ private module OutNodes {
...
-  predicate storeStepLibrary(Node pred, Content c, Node succ) {
-    // Complex flow summary
+  predicate summaryStoreStep(Node pred, Content c, Node succ) {
     exists(
-      ControlFlow::Node callCfn, CallableFlowSource source, AdjustedAccessPath sourceAp,
-      CallableFlowSink sink, AdjustedAccessPath sinkAp, boolean preservesValue,
-      LibraryCodeNodeState predState, AdjustedAccessPath ap
+      SourceDeclarationCallable sdc, CallableFlowSource source, AccessPath sourceAp,
+      CallableFlowSink sink, AccessPath sinkAp, boolean preservesValue,
+      SummaryInternalNodeState predState, AccessPath ap
```
Comment (tamasvajk) [inline csharp/ql/src/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll:1714]: as discuessed, `predAp` would probably be a better name for `ap`.

File: csharp/ql/src/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll [hunk 2]
```diff
@@ -1414,662 +1514,353 @@ private module OutNodes {
...
-  predicate readStepLibrary(Node pred, Content c, Node succ) {
-    // Complex flow summary
+  predicate summaryReadStep(Node pred, Content c, Node succ) {
     exists(
-      ControlFlow::Node callCfn, CallableFlowSource source, AdjustedAccessPath sourceAp,
-      CallableFlowSink sink, AdjustedAccessPath sinkAp, boolean preservesValue,
-      LibraryCodeNodeState succState, AdjustedAccessPath ap
+      SourceDeclarationCallable sdc, CallableFlowSource source, AccessPath sourceAp,
+      CallableFlowSink sink, AccessPath sinkAp, boolean preservesValue,
+      SummaryInternalNodeState succState, AccessPath ap
```
Comment (tamasvajk) [inline csharp/ql/src/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll:1741]: `succAp` instead of `ap`.

**Review by tamasvajk — CHANGES_REQUESTED** [review]

File: csharp/ql/src/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll [hunk 3]
```diff
@@ -1810,6 +1810,13 @@ module LibraryFlow {
       x.hasExprPath(source.getSource(call), entry.(ExprNode).getControlFlowNode(), _, callCfn)
     )
     or
+    // The source is an implicit array for a `params` parameter
+    exists(DataFlowCall call, int pos |
+      pos = source.(CallableFlowSourceArg).getArgumentIndex() and
+      entry.(ParamsArgumentNode).argumentOf(call, pos) and
+      callCfn = call.getControlFlowNode()
+    )
+    or
```
[the remark this answers is not in the document]
Comment (hvitved) [reply csharp/ql/src/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll]: We didn't have a test for this, I discovered it during an initial run of CSharp-Differences.

[the remark this answers is not in the document]
Comment (hvitved) [reply csharp/ql/src/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll]: I'll add a test.

File: csharp/ql/src/semmle/code/csharp/dataflow/internal/DataFlowDispatch.qll [hunk 4]
```diff
@@ -153,6 +166,7 @@ private module DispatchImpl {
           .(NonDelegateDataFlowCall)
           .getDispatchCall()
           .getADynamicTargetInCallContext(ctx.(NonDelegateDataFlowCall).getDispatchCall())
+          .getSourceDeclaration()
```
[the remark this answers is not in the document]
Comment (hvitved) [reply csharp/ql/src/semmle/code/csharp/dataflow/internal/DataFlowDispatch.qll:169]: Yes, a bug fix, but should only happen in rare cases.

File: csharp/ql/src/semmle/code/csharp/dataflow/internal/DelegateDataFlow.qll [hunk 5]
```diff
@@ -215,8 +186,7 @@ private predicate flowsFrom(
   or
   // Flow into a callable (delegate call)
   exists(
-    ExplicitParameterNode mid, CallContext prevLastCall, DelegateCall call, Callable c, Parameter p,
-    int i
+    ParameterNode mid, CallContext prevLastCall, DelegateCall call, Callable c, Parameter p, int i
```
[the remark this answers is not in the document]
Comment (hvitved) [reply csharp/ql/src/semmle/code/csharp/dataflow/internal/DelegateDataFlow.qll:189]: The reason why they are structured differently is to get a good join-order: the "flow into delegate" case is actually non-linear recursion (a conjunction with two recursive calls to `flowsFrom`).

**Review by tamasvajk — DISMISSED** [review]

**Comment by hvitved** [comment]
I was unable to reproduce the failing `library-tests/dataflow/collections/CollectionFlow.ql` test locally, so I have tried to rebase this PR to see if it makes a difference.

**Review by tamasvajk — APPROVED** [review]

**Pull request merged**